### ISSUE TITLE
Matlab script to generate test images in .mat format not available

### ISSUE THREAD

**Issue by ishmael1985** [twtygqyy/pytorch-vdsr#22]
For now, I'm using this from your LapSRN repo:

https://github.com/twtygqyy/pytorch-LapSRN/blob/master/Test/generate_test_mat.m

It would be nice if you can package this file together with your test images in your VDSR repo, so that users don't get mystified by the generation of your test data.

A few questions:

1) When converting to YUV space you normalized the pixel values, then splitting the channels, you multiply back the normalizing factor to get the mapped range in YUV space. Why the added normalization steps?
2) Why are the PSNR results so different if I try to reproduce your test script but using torchvision / PIL transforms? I'm getting results worse than bicubic for scaling factors 3 and 4 using torchvision / PIL, yet the Matlab script allows me to reproduce your reported results.
3) Relating to 2) how different is imresize() from Image.resize(resampling=PIL.Image.BICUBIC), and rgb2ycbcr() from Image.convert('YCbCr')? The possible differences are causing huge deviations in PSNR results, with PIL giving much better results than Matlab's equivalent calls when measuring bicubic interpolation performance.

**Comment by twtygqyy** [comment]
Hi @ishmael1985 thanks for the suggestion, I will make a push when I have time.
Regarding your questions, here are my answers:

1. Without normalization step, the pixel value range will be from 0~255, which is not friendly to CNN's training. (Takes much longer time to converge).

2. & 3. Because the implementation of PIL's bicubic interpolation is different from the one in Matlab. So far, only using the bicubic of Matlab could achieve the best PSNR score, because it has anti-aliasing function. Please take a look at this one and have a try. [link](https://www.reddit.com/r/MachineLearning/comments/6vdo51/p_matlab_bicubic_imresize_implemented_in_python/)

**Comment by ishmael1985** [comment]
@twtygqyy thanks for your effort but I'm still stumped with the differences. To make it clear, please first let me point out the code I've written to reproduce your results, and I'll try to highlight the important points:

https://github.com/ishmael1985/ml_research/tree/vdsr_pytorch

So I prepared the augmented training dataset of 291 images with this shell script (https://github.com/ishmael1985/ml_research/blob/vdsr_pytorch/vdsr_dataset.sh), which I wrote to emulate the output of your generate_train.m Matlab script. Important differences are that I used HDF5 gzip compression and that the shuffling of the 41x41 patches is done in batches of 512 images in order for me to save memory. It is obvious that I used Torchvision / PIL transforms to augment the dataset in contrast to using Matlab image processing toolbox, this is an important point.

To verify that my generated HDF5 file is good compared to the one generated by generate_train.m, I used a set of 91 images since my system can't handle the system memory requirements required for 291 images and generated HDF5 files using both scripts. Without HDF5 compression, the file sizes are similar (around 2 GB for each if I remember right) and the properties are pretty much the same when viewing both through HDFView.

The trained model I got (https://github.com/ishmael1985/ml_research/blob/vdsr_pytorch/checkpoint/model_epoch_50.pth) checks out with your eval.py and I get the following output.

```
(tensorflow) D:\ishmael\pytorch-vdsr>python eval.py --cuda --model=D:\ishmael\work\ml
_research\checkpoint\model_epoch_50.pth
=> use gpu id: '0'
Processing  Set5_mat\baby_GT_x2.mat
Processing  Set5_mat\bird_GT_x2.mat
Processing  Set5_mat\butterfly_GT_x2.mat
Processing  Set5_mat\head_GT_x2.mat
Processing  Set5_mat\woman_GT_x2.mat
Scale= 2
Dataset= Set5
PSNR_predicted= 37.476038452066405
PSNR_bicubic= 33.691134694590936
It takes average 0.30576062202453613s for processing
Processing  Set5_mat\baby_GT_x3.mat
Processing  Set5_mat\bird_GT_x3.mat
Processing  Set5_mat\butterfly_GT_x3.mat
Processing  Set5_mat\head_GT_x3.mat
Processing  Set5_mat\woman_GT_x3.mat
Scale= 3
Dataset= Set5
PSNR_predicted= 33.5398275220277
PSNR_bicubic= 30.409157826579882
It takes average 0.024960136413574217s for processing
Processing  Set5_mat\baby_GT_x4.mat
Processing  Set5_mat\bird_GT_x4.mat
Processing  Set5_mat\butterfly_GT_x4.mat
Processing  Set5_mat\head_GT_x4.mat
Processing  Set5_mat\woman_GT_x4.mat
Scale= 4
Dataset= Set5
PSNR_predicted= 30.944941757610604
PSNR_bicubic= 28.433681495312406
It takes average 0.012480068206787109s for processing
```

It is lower than your reported results but for now I'll accept that. What I don't understand is when I try to load the same model to my evaluation script (https://github.com/ishmael1985/ml_research/blob/vdsr_pytorch/sr_test.py), and I get completely different results!

```
(tensorflow) D:\ishmael\work\ml_research>python sr_test.py --image_folder=D:\ish
mael\work\SRCNN-Tensorflow\Test\Set5 --sample_size=-1 --model=checkpoint\model_e
poch_50.pth --scale 2 3 4 --cuda
C:\Users\Ishmael\Anaconda3\envs\tensorflow\lib\site-packages\h5py\__init__.py:36
: FutureWarning: Conversion of the second argument of issubdtype from `float` to
 `np.floating` is deprecated. In future, it will be treated as `np.float64 == np
.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Upscale factor =  2
Average PSNR for 5 samples = 36.0077720393054
Average bicubic PSNR for 5 samples = 35.01241467898105
Upscale factor =  3
Average PSNR for 5 samples = 32.31447073860615
Average bicubic PSNR for 5 samples = 33.256938525238944
Upscale factor =  4
Average PSNR for 5 samples = 29.911576289426772
Average bicubic PSNR for 5 samples = 32.258016213377154
```

Bicubic PSNR is doing better for me than the trained model! And it's higher by quite a difference compared to the one shown by the eval.py script. It's worth mentioning the PSNR values I'm getting with my trained model is considerably lower than the one I got using your eval.py script.

To confirm whether my results are reliable, I used your trained model (https://github.com/twtygqyy/pytorch-vdsr/blob/master/model/model_epoch_50.pth) with both scripts and I am getting pretty much the same results.

I've re-checked the implementation for sr_test.py script I wrote and I can't figure out what I could have possibly done wrong. I used Torchvision's Resize and CenterCrop transforms to get the interpolated images and then measure the PSNR between the interpolated image and the ground truth. This contrasts to the pre-processed interpolated images stored in .mat format and using border shaving by the eval.py script.

I appreciate it if you've read this far. In that case, what could I have possibly missed in my sr_test.py evaluation script?

**Comment by ishmael1985** [comment]
I should also mention that I have used the Python version of imresize and that I'm still getting more or less the same results as I reported in my previous comment.

**Comment by twtygqyy** [comment]
Hi @ishmael1985 I guess there might be some problems with your sr_test code. If you use Python imresize for both training and inference, although the PSNR score would not be as good as the one by Matlab, it should still much higher than bicubic.
I haven't checked your code in detail, but please check  the two images for PSNR calculation function (should be 0~255 in range, better in uint8 type). Please also visualize the results, which makes it easier to find the problem.
I will take a look at your code in the following days when time allows.

**Comment by ishmael1985** [comment]
Thanks a lot for the hints. I changed the sr_test.py code accordingly (https://github.com/ishmael1985/ml_research/commit/0b4387fc65b4270fbe4525872e2951d6a0ae0e99) and also compared the output, interpolated and ground truth images. Subjectively, the results look like they are expected.
I switched to using sci-kit image library to perform the normalization and compute the psnr. However, my results don't seem to be as good as eval.py. This is what I got for my trained model:

```
(tensorflow) D:\ishmael\work\ml_research>python sr_test.py --image_folder=D:\ish
mael\work\SRCNN-Tensorflow\Test\Set5 --sample_size=-1 --model=checkpoint\model_e
poch_50.pth --scale 2 3 4 --cuda
C:\Users\Ishmael\Anaconda3\envs\tensorflow\lib\site-packages\h5py\__init__.py:36
: FutureWarning: Conversion of the second argument of issubdtype from `float` to
 `np.floating` is deprecated. In future, it will be treated as `np.float64 == np
.dtype(float).type`.
  from ._conv import register_converters as _register_converters
C:\Users\Ishmael\Anaconda3\envs\tensorflow\lib\site-packages\skimage\util\dtype.
py:122: UserWarning: Possible precision loss when converting from float32 to uin
t8
  .format(dtypeobj_in, dtypeobj_out))
Upscale factor =  2
Average PSNR for 5 samples = 35.984107860246155
Average bicubic PSNR for 5 samples = 32.32890705899756
Upscale factor =  3
Average PSNR for 5 samples = 32.30503052606939
Average bicubic PSNR for 5 samples = 29.07729627941132
Upscale factor =  4
Average PSNR for 5 samples = 29.905922810749967
Average bicubic PSNR for 5 samples = 27.102245119910997
```

With the pre-trained model you provided, it doesn't seem to get that much better:

```
 (tensorflow) D:\ishmael\work\ml_research>python sr_test.py --image_folder=D:\ish
mael\work\SRCNN-Tensorflow\Test\Set5 --sample_size=-1 --model=D:\ishmael\pytorch
-vdsr\model\model_epoch_50.pth --scale 2 3 4 --cuda
C:\Users\Ishmael\Anaconda3\envs\tensorflow\lib\site-packages\h5py\__init__.py:36
: FutureWarning: Conversion of the second argument of issubdtype from `float` to
 `np.floating` is deprecated. In future, it will be treated as `np.float64 == np
.dtype(float).type`.
  from ._conv import register_converters as _register_converters
C:\Users\Ishmael\Anaconda3\envs\tensorflow\lib\site-packages\torch\serialization
.py:325: SourceChangeWarning: source code of class 'torch.nn.modules.conv.Conv2d
' has changed. you can retrieve the original source code by accessing the object
's source attribute or set `torch.nn.Module.dump_patches = True` and use the pat
ch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
C:\Users\Ishmael\Anaconda3\envs\tensorflow\lib\site-packages\skimage\util\dtype.
py:122: UserWarning: Possible precision loss when converting from float32 to uin
t8
  .format(dtypeobj_in, dtypeobj_out))
Upscale factor =  2
Average PSNR for 5 samples = 36.004927808054575
Average bicubic PSNR for 5 samples = 32.32890705899756
Upscale factor =  3
Average PSNR for 5 samples = 32.264724336405244
Average bicubic PSNR for 5 samples = 29.07729627941132
Upscale factor =  4
Average PSNR for 5 samples = 29.500868156032244
Average bicubic PSNR for 5 samples = 27.102245119910997
```

Even with the Python version of imresize that you have pointed out, I still haven't managed to reproduce your results. I may have overlooked a little detail since I used Torchvision's CenterCrop, nevertheless, thanks for your help!

**Comment by ishmael1985** [comment]
I'm finally getting close enough results due to changing the YCbCr colorspace conversion routine, from this commit: https://github.com/ishmael1985/ml_research/commit/9ec54934947985355423237e203cb77fad2fae33

The result is still a little lower, perhaps, as you recommended, I should retrain the model.

```
(tensorflow) D:\ishmael\work\ml_research>python sr_test.py --image_folder=D:\ish
mael\work\SRCNN-Tensorflow\Test\Set5 --sample_size=-1 --model=checkpoint\model_e
poch_50.pth --scale 2 3 4 --cuda
C:\Users\Ishmael\Anaconda3\envs\tensorflow\lib\site-packages\h5py\__init__.py:36
: FutureWarning: Conversion of the second argument of issubdtype from `float` to
 `np.floating` is deprecated. In future, it will be treated as `np.float64 == np
.dtype(float).type`.
  from ._conv import register_converters as _register_converters
C:\Users\Ishmael\Anaconda3\envs\tensorflow\lib\site-packages\skimage\util\dtype.
py:122: UserWarning: Possible precision loss when converting from float32 to uin
t8
  .format(dtypeobj_in, dtypeobj_out))
Upscale factor =  2
Average PSNR for 5 samples = 37.24455212084543
Average bicubic PSNR for 5 samples = 33.63790475425825
Upscale factor =  3
Average PSNR for 5 samples = 33.61805743157027
Average bicubic PSNR for 5 samples = 30.39497394219554
Upscale factor =  4
Average PSNR for 5 samples = 31.200405881893516
Average bicubic PSNR for 5 samples = 28.41963087997638
```

**Comment by ishmael1985** [comment]
Seen that you have uploaded this (https://github.com/twtygqyy/pytorch-vdsr/blob/master/data/generate_test_mat.m), I am closing this issue.

**Comment by twtygqyy** [comment]
@ishmael1985 Glad to hear that you found the solution, there are lot of tricks for calculating PNSR value and the scores are very sensitive. Personally, I also find using Matlab quite annoying, while it is still the only choice so far if you want to reproduce the results of state-of-the-art papers. 
Please let me know if you have some new findings, and I'm happy to share my experience on this topic.
